feat: flagd in-process provider (#633)
open-feature/js-sdk-contrib: libs/providers/flagd/README.md

## Installation

### npm

```
$ npm install @openfeature/flagd-provider
```

### yarn

```

> # yarn require manual installation of peer dependencies
> $ npm install @openfeature/server-sdk @grpc/grpc-js @openfeature/flagd-core

Reviewer: ```suggestion
# yarn requires manual installation of peer dependencies
$ yarn install @openfeature/server-sdk @grpc/grpc-js @openfeature/flagd-core
```

> $ yarn install @openfeature/server-sdk @grpc/grpc-js @openfeature/flagd-core

Reviewer: ```suggestion
$ yarn add @openfeature/server-sdk @grpc/grpc-js @openfeature/flagd-core
```

### Available Configuration Options

| Option name           | Environment variable name      | Type    | Default   | Supported values |
|-----------------------|--------------------------------|---------|-----------|------------------|
| host                  | FLAGD_HOST                     | string  | localhost |                  |
| port                  | FLAGD_PORT                     | number  | 8013      |                  |
| tls                   | FLAGD_TLS                      | boolean | false     |                  |
| socketPath            | FLAGD_SOCKET_PATH              | string  | -         |                  |
| resolverType          | FLAGD_SOURCE_RESOLVER          | string  | rpc       | rpc, in-process  |
| selector              | FLAGD_SOURCE_SELECTOR          | string  | -         |                  |
| cache                 | FLAGD_CACHE                    | string  | lru       | lru,disabled     |
| maxCacheSize          | FLAGD_MAX_CACHE_SIZE           | int     | 1000      |                  |
| maxEventStreamRetries | FLAGD_MAX_EVENT_STREAM_RETRIES | int     | 5         |                  |

> Given below are examples on usage patterns.

Reviewer 2: ```suggestion
Below are examples of usage patterns.
```

### Remote flag resolving over RPC

This is the default mode of operation of the provider.
In this mode, FlagdProvider communicates with flagd via the gRPC protocol.
Flag evaluations take place remotely at the connected [flagd](https://flagd.dev/) instance.

> ```

Reviewer 2: ```suggestion
```ts
```

  OpenFeature.setProvider(new FlagdProvider())
```

> In the above example, provider expect flagd to be available at `localhost:8013` (default host and port).

Reviewer 2: ```suggestion
In the above example, the provider expects flagd to be available at `localhost:8013` (default host and port).
```

> Alternatively, you can use socket paths to connect to flagd as given below,

Reviewer 2: ```suggestion
Alternatively, you can use socket paths to connect to flagd.
```

```
  OpenFeature.setProvider(new FlagdProvider({
      socketPath: "/tmp/flagd.socks",
  }))
```

### In-process resolver

This mode performs flag evaluations locally (in-process).
Flag configurations for evaluation are obtained via gRPC protocol using [sync protobuf schema](https://buf.build/open-feature/flagd/file/main:sync/v1/sync_service.proto) service definition.

```
  OpenFeature.setProvider(new FlagdProvider({
      resolverType: 'in-process',
  }))
```

> In the above example, provider expect flag sync service implementation to be available at `localhost:8013` (default host and port).

Reviewer 2: ```suggestion
In the above example, the provider expects a flag sync service implementation to be available at `localhost:8013` (default host and port).
```

Required peer dependencies

```

> $ npm install @openfeature/server-sdk @grpc/grpc-js @openfeature/flagd-core

Author: Yup, saw the conversation on npm vs yarm strategy. I will update this section

> $ npm install @openfeature/server-sdk @grpc/grpc-js @openfeature/flagd-core

Author: Updated with the latest commit

  }))
```

### Remote flag resolving using Unix Socket

```
  OpenFeature.setProvider(new FlagdProvider({
      socketPath: "/tmp/flagd.socks",
  }))
```

> ### In-process resolver with gRPC sync connectivity

Reviewer: Add a link [here](https://flagd.dev/concepts/syncs/#grpc-sync) so people understand what this is.

Author: I updated the readme with feedback. You can check the latest here - https://github.com/open-feature/js-sdk-contrib/blob/454e08bbdd1fd94ca5ed19b97050b24f3a691ec1/libs/providers/flagd/README.md 

Regarding the link, this usage is NOT flagd's sync compatibility on grpc. So I think the proposed link can mislead others. I don't think flagd docs have a dedicated section on sync contract that can be used by an in-process provider .

Reviewer: What if we just linked to https://buf.build/open-feature/flagd/docs/main:flagd.sync.v1 ?

Author: I have a linked section in the latest readme. Will this be enough ? :) 

> Flag configurations for evaluation are obtained via gRPC protocol using [sync protobuf schema](https://buf.build/open-feature/flagd/file/main:sync/v1/sync_service.proto) service definition.`

Reviewer: yep

```
  OpenFeature.setProvider(new FlagdProvider({
      host: 'localhost',

>       port: 8013,

Reviewer 2: This may imply that the in-process provider works with flagd. Should we use the document the flagd proxy port? Perhaps we should also include a note underneither explaining where the configuration is expected to come from.

Reviewer: I agree we should change the port in this example, but I think more importantly, we should put explanation and links to [this](https://flagd.dev/concepts/syncs/#grpc-sync) in the js-docs so people can be pointed in the right direction from the code.

Author: Good point, will update the docs and usage example
